INDDEX reports - final indicators in data source

esoergel opened the pull request:
https://docs.google.com/document/d/1fy7_JYhZwy4z8VA8ikygXSXTTSHNqaTFUWaj1AHcVD0/edit

##### SUMMARY
Picking up from https://github.com/dimagi/commcare-hq/pull/27036
This implements "gap" information (defined [here](https://docs.google.com/document/d/1FeOQ63GaZyOeU1USjBgWhLn9wSBfOQoUiCo1lj5X51U/edit#heading=h.t2ckxl3uaivu))

##### FEATURE FLAG
<!--- If this is specific to a feature flag, which one? -->

##### PRODUCT DESCRIPTION
<!--- For non-invisible changes, describe user-facing effects. -->

esoergel committed:
Set conversion factor gap columns

esoergel committed:
Calculate FCT gaps

esoergel commented on custom/inddex/food.py:
```diff
@@ -195,7 +231,30 @@ def _set_composition(self):
+            }[self.fct_data_used]
+
+        if self.is_recipe:
+            if all(i.fct_gap_code == FctGaps.AVAILABLE for i in ingredients):
```
unfortunately, calculating a gap code for a recipe requires context of the ingredients in that recipe, so I'm setting it during the `enrich` phase.

kaapstorm commented on custom/inddex/food.py:
```diff
@@ -210,6 +231,28 @@ def _set_composition(self):
+                'reference_food_code': FctGaps.REFERENCE,
+            }[self.fct_data_used]
+
+        if self.is_recipe:
+            if all(i.fct_gap_code == FctGaps.AVAILABLE for i in ingredients):
```
I'm having difficulty wrapping my head around the difference in behavior between standard recipes and non-standard recipes. (Maybe an example of a non-standard recipe would help. Maybe [horse-flesh-flavored ice-cream](http://food.thecookbk.com/know-horse-flesh-flavored-ice-cream-japan/)? That sounds non-standard.)

In both cases `self.is_recipe` is `True` ... but in `enrich_rows()` is it possible for `recipe_possibilities` to have more than one item? So `recipe` would be `None` and `row.set_fct_gap()` would be called with `ingredients=None`. But `row.is_recipe` would be `True` for at least two rows, and line 249 would error out with a `TypeError`.

If it should not be possible, maybe add an `assert`? Or treat `ingredients` as ingredients of both recipes??? 

...

(Very nitpicky suggestion: In `enrich_rows()` you can move `ingredients` into the `else` clause of `if not recipe` because it only gets used there ... *ooorrr* you can do this neat trick:
```python
        ingredients, recipe_possibilities = [], []
        for row in rows:
            (ingredients, recipe_possibilities)[row.is_recipe].append(row)
```
:stuck_out_tongue_winking_eye: (OK, no, maybe don't do that.))

esoergel replied:
> horse-flesh-flavored ice-cream
:fearful: 

Excellent question - I should probably add some comments describing this.  Basically a "standard recipe" is one that's defined in their lookup tables.  The user says "I made gumbo", and we know that that means X amount of Okra, a cube of Maggi, some fish powder, and so on.  When this is the case, these foods are inferred based on the recipe, and this report inserts rows based on lookup table data.  With nonstandard recipes, the user defines the ingredients in the recipe, each of which is a new food item.  Those rows appear in the report based on rows in the food case UCR.

> In `enrich_rows()` is it possible for `recipe_possibilities` to have more than one item? 

I don't _think_ so (well, with one exception I'll come back to).  My intent there was to give up if that situation arises and treat each row as a standalone food item, just like with the `NO_RECIPE` rows.

> line 249 would error out with a TypeError

Good catch, thank you!  I think better behavior would be to change that to `if recipe and ingredients`, so that if `not ingredients`, then the gap code would be `FctGaps.NOT_AVAILABLE`.

Ha!  That's a neat trick, although yes, perhaps too esoteric to use.  I was actually thinking it'd be nice to have something to split an array, like [clojure's `partition-by`](https://clojuredocs.org/clojure.core/partition-by) (apparently [lodash has one too](https://lodash.com/docs/4.17.15#partition)).  I had `ingredients` defined up there in an attempt to illustrate that this split the lists, but agree that it should probably be moved, given the current state of the code.

--------------------------------

The exception I mentioned is related to this:
> Or treat `ingredients` as ingredients of both recipes???

This is complication that I don't think is currently well-handled.  Sometimes people make a big batch of food and eat it in more than one sitting. :man_shrugging:  The way this manifests is that the first time the recipe is referenced, all ingredients are noted.  Subsequently, they just refer back to that first recipe via the `already_reported_recipe_case_id` property.  At the moment I'm handling that poorly, mostly because I didn't fully understand how it worked when I wrote that part.  However it _is_ consistent with the examples, so I just think the spec doesn't really know how to deal with that properly.  I proposed duplicating the ingredient-based indicators to both recipes yesterday, and Nick will reach out for clarification on that point.

This conversation is also making me realize that the same problem can exist with standard recipes too - as-is, I think the report will insert all ingredients of a standard recipe every time the standard recipe is mentioned, which also doesn't seem ideal (unless the app references the first reference to a standard recipe via `already_reported_recipe_case_id`?).

kaapstorm replied:
>  like clojure's partition-by 

Ooo! That's [`itertools.groupby`](https://docs.python.org/3/library/itertools.html#itertools.groupby):
```python
In [62]: one_to_five = (i + 1 for i in range(5))
In [63]: [list(items) for group_val, items in groupby(one_to_five, lambda x: bool(x % 3))]
Out[63]: [[1, 2], [3], [4, 5]]
```

Also pretty cute is [`itertools.tee`](https://docs.python.org/3/library/itertools.html#itertools.tee).

kaapstorm replied:
Also, I love itertools.

esoergel replied:
I actually looked at that one!  Unfortunately it only groups consecutive values.  Here's an attempt to split a list of numbers into evens and odds:

```python
In [1]: from itertools import groupby

In [2]: [list(items) for _, items in groupby([1, 2, 4, 6, 7], lambda x: x % 2)]
Out[2]: [[1], [2, 4, 6], [7]]
```
I read [this Ned Batchelder post](https://nedbatchelder.com/blog/201306/filter_a_list_into_two_parts.html) on the problem and concluded that it wasn't worth the vebosity.  Although the solution he proposes does use `itertools.tee`, actually.

esoergel committed:
Manage recipe/ingredient edge cases better

`already_reported_recipe_case_id` is only set on recipes, so it doesn't make
sense to have that there.
Treat already-reported recipes as recipes with no ingredients (for now)
If a recipe shows up with no ingredients, mark it NOT_AVAILABLE
Use `uuid` to filter `recipe` out of rows to ensure nothing else is dropped
(This last one shouldn't change anything, as anything that `is_recipe`
can't appear in `ingredients` anyways, as its `recipe_id` is its own
`case_id`)

kaapstorm reviewed (approved):
Sweet, thank you.

esoergel committed:
Add docstring

esoergel committed:
Forgot to update the test data with that last change

kaapstorm reviewed (approved):
:heart_eyes: awesome docstring
